server/src/main/java/org/apache/druid/discovery/WorkerNodeService.java

```diff
@@ -33,16 +33,19 @@
   private final String ip;
   private final int capacity;
   private final String version;
+  private final String tier;
 
   public WorkerNodeService(
       @JsonProperty("ip") String ip,
       @JsonProperty("capacity") int capacity,
-      @JsonProperty("version") String version
+      @JsonProperty("version") String version,
+      @JsonProperty("tier") String tier
   )
   {
     this.ip = ip;
     this.capacity = capacity;
     this.version = version;
+    this.tier = tier;
```

Comment:
Many Druid users follow rolling upgrade (as described in http://druid.io/docs/latest/operations/rolling-updates.html ) . If you only upgrade overlord  then, at overlord, tier is going to be null which would lead to a NPE in `WorkerSelectUtils.getTierWorkers(..)` at line "Maps.filterValues(workerMap, workerInfo -> workerInfo.getWorker().getTier().equals(tier))" and maybe in some other places as well.
 that should probably be fixed or something needs to go in release nodes along the lines of making sure that new tier based strategies are only to be used when there has been at least one restart of overlord after all middleManager upgrade (due to second cluster upgrade or forced restart of overlord after MM upgrade or whatever)

Reply:
If users want to use this feature, they must upgrade overlord and all middleManager ( follow the rolling upgrade is ok). I have noted this in the document. An extra restart of overlord is not necessary.